js/hang-ui/src/Components/publish/PublishUIContextProvider.tsx

```diff
@@ -0,0 +1,229 @@
+import type HangPublish from '@kixelated/hang/publish/element';
+import type { HangPublishInstance } from '@kixelated/hang/publish/element';
+import type { JSX } from 'solid-js';
+import { createContext, createEffect, createSignal } from 'solid-js';
+
+type PublishStatus =
+    | 'no-url'
+    | 'disconnected'
+    | 'connecting'
+    | 'live'
+    | 'audio-only'
+    | 'video-only'
+    | 'select-source';
+
+type PublishUIContextValue = {
+    hangPublish: () => HangPublish | undefined;
+    cameraDevices: () => MediaDeviceInfo[];
+    microphoneDevices: () => MediaDeviceInfo[];
+    publishStatus: () => PublishStatus;
+    microphoneActive: () => boolean;
+    cameraActive?: () => boolean;
+    screenActive?: () => boolean;
+    fileActive?: () => boolean;
+    nothingActive?: () => boolean;
+    selectedCameraSource?: () => MediaDeviceInfo['deviceId'] | undefined;
+    selectedMicrophoneSource?: () => MediaDeviceInfo['deviceId'] | undefined;
+    setFile: (file: File) => void;
+};
+
+type PublishUIContextProviderProps = {
+    hangPublish: () => HangPublish | undefined;
+    children: JSX.Element;
+};
+
+export const PublishUIContext = createContext<PublishUIContextValue>();
+
+export default function PublishUIContextProvider(
+    props: PublishUIContextProviderProps
+) {
+    const [cameraDevices, setCameraMediaDevices] = createSignal<
+        MediaDeviceInfo[]
+    >([]);
+    const [selectedCameraSource, setSelectedCameraSource] = createSignal<
+        MediaDeviceInfo['deviceId'] | undefined
+    >();
+    const [microphoneDevices, setMicrophoneMediaDevices] = createSignal<
+        MediaDeviceInfo[]
+    >([]);
+    const [selectedMicrophoneSource, setSelectedMicrophoneSource] =
+        createSignal<MediaDeviceInfo['deviceId'] | undefined>();
+    const [cameraActive, setCameraActive] = createSignal<boolean>(false);
+    const [screenActive, setScreenActive] = createSignal<boolean>(false);
+    const [microphoneActive, setMicrophoneActive] =
+        createSignal<boolean>(false);
+    const [fileActive, setFileActive] = createSignal<boolean>(false);
+    const [nothingActive, setNothingActive] = createSignal<boolean>(false);
+    const [publishStatus, setPublishStatus] =
+        createSignal<PublishStatus>('no-url');
+
+    const setFile = (file: File) => {
+        const hangPublishEl = props.hangPublish();
+        if (!hangPublishEl) return;
+
+        hangPublishEl.file = file;
+        hangPublishEl.source = 'file';
+        hangPublishEl.video = true;
+        hangPublishEl.audio = true;
+        setFileActive(true);
+    };
+
+    const value: PublishUIContextValue = {
+        hangPublish: props.hangPublish,
+        cameraDevices,
+        microphoneDevices,
+        publishStatus,
+        cameraActive,
+        screenActive,
+        microphoneActive,
+        fileActive,
```

Comment:
I'm open to making these on property - `activeSources`, and then having the individual buttons check for membership in the Set/array.